common/xPowerManager.h

```diff
@@ -0,0 +1,769 @@
+                }
+            #endif
+        #ifdef DIAGNOSE_POWER
+            PrintPowerState(newState);
+        #endif // DIAGNOSE_POWER
+        }
+    } xpd_sd;    // one object must exist
+
+
+    class PowerDomain_CPU : public xPowerDomain {
+    public:            
+        PowerDomain_CPU() : xPowerDomain() {}    // Link domain to powerDomains
+        PDType id() override { return pwr4_CPU; } // define ID
+        const char* name() override { return "CPU"; }  // define name
+        uint32_t timeout() override { return PWRMAN_CPUTIMEOUT; }  // Override timeout
+        
+        void SetPower(bool newState) {  // Nothing to do, CPU sleep mode is handled by xPowerManager
+            // SaberBase::motion_active_ = newState;   // activate / deactivate motion
+            // gyroscope->Loop();                      // run loop once to allow sensor state machine to register the change
+        #ifdef DIAGNOSE_POWER
+            PrintPowerState(newState);
+        #endif // DIAGNOSE_POWER                
+        }   
+    } xpd_cpu;    // one object must exist
+
+
+    // POWER SUBSCRIBERS
+    // ==================================================================================
+    // Power subscriber objects must inherit from this class 
+    class xPowerSubscriber {
+    friend class xPowerManager;
+    private:
+        PDType_base subscribedDomains;  // binary map
+        xPowerSubscriber* next;
+    protected:
+        virtual const char* name() {}       // Subscriber's name, defined by child 
+        virtual bool HoldPower() { return false; }   // Override to pause timeout
+        virtual void PwrOn_Callback() {}          // Subscriber-specific code to be executed when subscriber power goes ON
+        virtual void PwrOff_Callback() {}         // Subscriber-specific code to be executed when subscriber power goes OFF
+
+    public:    
+        // Constructor lists subscriber in powerman.subscribers
+        xPowerSubscriber(PDType_base domains) { 
+            if (!domains) return;
+            subscribedDomains = domains;        // mark domains
+            next = powerman.subscribers;        // link to subscribers list
+            powerman.subscribers = this;
+        }                      
+                                
+        // Check if all the subscribed domains are active
+        bool IsOn() {  return ((powerman.powerState & subscribedDomains) == subscribedDomains); }
+
+        // Keep power ON until timeout on all subscribed domains. Turns domains ON if needed.
+        // Returns true if subscriber went on just now
+        // If specified, timeout_ must point to a vector of uint32_t with the tineout value in millis for each subscribed domain, e.g. 'RequestPower(&(const uint32_t t=100));' or 'RequestPower(&(const uint32_t[] t={100,200}));'
+        // If not specified, domain-specific timeout will be applied to each subscribed domain
+        bool RequestPower(uint32_t* timeout_ = 0) {
```

Comment:
Why do you have to be a subscriber to request power from a domain?
It seems like several places in the code inherit from xPowerSubscriber just to keep things alive, but it could just as easily be done with a call like:

  powerman.RequestPower(pwr4_Booster);

Also, the way to specify timeouts seems really weird, and is only ever used with a single timeout, so why not just keep it simple and have *one* timeout?

Finally, code style is to use ending underscore for class variables, not local variables.

Reply:
Without subscribers as distinct objects we would loose the callbacks and HoldPower functionality.

The weird specification is what allows a subscriber to request different timeouts for different domains. It's for future extension, feel free to remove it if you don't find any use for it.

Reply:
I'm not saying we shouldn't have subscribers, it's a useful function.
I'm saying that some classes don't need it, because all they want is to keep some power domain on.

Reply:
> The weird specification is what allows a subscriber to request different timeouts for different domains. It's for future 
> extension, feel free to remove it if you don't find any use for it.

I don't want to check in things that we don't need unless they are implemented in such a way that they don't use any flash memory.  (Adding a function/method that we never call is fine, but a function with arguments that we never use is not.)